Source: nodejs-mobile/nodejs-mobile@de4c40ab8c4450c4641ce78a83cae5657a1fa286
Source: nodejs-mobile/nodejs-mobile@c26a8fd84877ba4dd0832e2619dd31e8d5839070
Source: nodejs-mobile/nodejs-mobile@a8a59e78c5dec30c37fffbe3d582c4edd3830c5f

diff --git a/lib/internal/modules/cjs/loader.js b/lib/internal/modules/cjs/loader.js
index f9c4a62f023a5e8e635079b383fda74774377b8b..19f23183807b432b3d7b3f80f18e6d849246f243 100644
--- a/lib/internal/modules/cjs/loader.js
+++ b/lib/internal/modules/cjs/loader.js
@@ -2216,7 +2216,8 @@ Module.createRequire = createRequire;
  */
 Module._initPaths = function() {
   const homeDir = isWindows ? process.env.USERPROFILE : safeGetenv('HOME');
-  const nodePath = isWindows ? process.env.NODE_PATH : safeGetenv('NODE_PATH');
+  const nodePath = isWindows || process.platform === 'android' ?
+    process.env.NODE_PATH : safeGetenv('NODE_PATH');
 
   // process.execPath is $PREFIX/bin/node except on Windows where it is
   // $PREFIX\node.exe where $PREFIX is the root of the Node.js installation.
diff --git a/src/node_credentials.cc b/src/node_credentials.cc
index d1f7c6e2a801fee3bbcbc5ac2a31cf6781c9f105..2ffaaaca5289a7c7389414fb95d54c17583cb91a 100644
--- a/src/node_credentials.cc
+++ b/src/node_credentials.cc
@@ -286,6 +286,8 @@ static void GetEGid(const FunctionCallbackInfo<Value>& args) {
   args.GetReturnValue().Set(static_cast<uint32_t>(getegid()));
 }
 
+#ifndef __ANDROID__
+// Android blocks setting Uid and Gid.
 static void SetGid(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
   CHECK(env->owns_process_state());
@@ -369,6 +371,7 @@ static void SetEUid(const FunctionCallbackInfo<Value>& args) {
     args.GetReturnValue().Set(0);
   }
 }
+#endif  // ifndef __ANDROID__
 
 static void GetGroups(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
@@ -392,6 +395,8 @@ static void GetGroups(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+#ifndef __ANDROID__
+// Android blocks setting groups
 static void SetGroups(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
 
@@ -427,6 +432,7 @@ static void SetGroups(const FunctionCallbackInfo<Value>& args) {
 
   args.GetReturnValue().Set(0);
 }
+#endif  // ifndef __ANDROID__
 
 static void InitGroups(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
@@ -486,11 +492,13 @@ void RegisterExternalReferences(ExternalReferenceRegistry* registry) {
   registry->Register(GetGroups);
 
   registry->Register(InitGroups);
+#ifndef __ANDROID__
   registry->Register(SetEGid);
   registry->Register(SetEUid);
   registry->Register(SetGid);
   registry->Register(SetUid);
   registry->Register(SetGroups);
+#endif
 #endif  // NODE_IMPLEMENTS_POSIX_CREDENTIALS
 }
 
@@ -514,11 +522,13 @@ static void Initialize(Local<Object> target,
 
   if (env->owns_process_state()) {
     SetMethod(context, target, "initgroups", InitGroups);
+#ifndef __ANDROID__
     SetMethod(context, target, "setegid", SetEGid);
     SetMethod(context, target, "seteuid", SetEUid);
     SetMethod(context, target, "setgid", SetGid);
     SetMethod(context, target, "setuid", SetUid);
     SetMethod(context, target, "setgroups", SetGroups);
+#endif
   }
 #endif  // NODE_IMPLEMENTS_POSIX_CREDENTIALS
 }
diff --git a/src/node_internals.h b/src/node_internals.h
index bb45d73e909071427498ed10fb2186743cfa5a49..3d6deb66259f174c4a26884059793cd2b3d12923 100644
--- a/src/node_internals.h
+++ b/src/node_internals.h
@@ -327,7 +327,7 @@ class ThreadPoolWork {
 
 // Functions defined in node.cc that are exposed via the bootstrapper object
 
-#if defined(__POSIX__) && !defined(__ANDROID__) && !defined(__CloudABI__)
+#if defined(__POSIX__) && !(defined(__ANDROID__) && __ANDROID_API__ < 21) && !defined(__CloudABI__)  // NOLINT(whitespace/line_length)
 #define NODE_IMPLEMENTS_POSIX_CREDENTIALS 1
 #endif  // defined(__POSIX__) && !defined(__ANDROID__) && !defined(__CloudABI__)
 
